File: src/lib/persist/io/io-queue.ts
import { IOOperationType, ReadIOOperation } from "../../globals"
import IOOperation from "./io-operation"
import WriteIOOperation from "./write-io-operation"

export default class IOQueue {
    readQueue: ReadIOOperation[] = []
    writeQueue: WriteIOOperation[] = []

    constructor() {}

    getReady(): [ReadIOOperation[], WriteIOOperation[]] {
        if (!this.opPending()) {
            return [[], []]
        }
        const readOps = this.readQueue
        const writeOps = this.writeQueue
        this.readQueue = []
        this.writeQueue = []
        for (const op of readOps) {
            op.processing = true
        }
        for (const op of writeOps) {
            op.processing = true
        }
        return [readOps, writeOps]
    }

    drain(): [ReadIOOperation[], WriteIOOperation[]] {
        const readOps = this.readQueue
        const writeOps = this.writeQueue
        this.readQueue = []
        this.writeQueue = []
        return [readOps, writeOps]
    }

    /**
     * consolidate all items ready for processing to oldQueue and return true if there are any
     */
    opPending(): boolean {
        return this.readQueue.length > 0 || this.writeQueue.length > 0
    }

    enqueue(op: IOOperation) {
        if (op.op === IOOperationType.WRITE) {
            this.writeQueue.push(op as WriteIOOperation)
        } else {
            this.readQueue.push(op as ReadIOOperation)
        }
    }
}

File: external/logsrdcpp/src/persist/io/IOQueue.cpp
#include "IOQueue.h"

namespace logsrd {

void IOQueue::enqueue(IOOperation* op) {
    if (op->opType == IOOperationType::WRITE) {
        writeQueue_.push_back(op);
    } else {
        readQueue_.push_back(op);
    }
}

std::pair<std::span<IOOperation*>, std::span<IOOperation*>> IOQueue::getReady() {
    // Mark all ops as processing
    for (auto* op : readQueue_) op->processing = true;
    for (auto* op : writeQueue_) op->processing = true;

    auto result = std::pair(
        std::span<IOOperation*>(readQueue_.data(), readQueue_.size()),
        std::span<IOOperation*>(writeQueue_.data(), writeQueue_.size())
    );

    readQueue_.clear();
    writeQueue_.clear();
    return result;
}

std::pair<std::span<IOOperation*>, std::span<IOOperation*>> IOQueue::drain() {
    auto result = std::pair(
        std::span<IOOperation*>(readQueue_.data(), readQueue_.size()),
        std::span<IOOperation*>(writeQueue_.data(), writeQueue_.size())
    );
    readQueue_.clear();
    writeQueue_.clear();
    return result;
}

bool IOQueue::opPending() const {
    return !readQueue_.empty() || !writeQueue_.empty();
}

} // namespace logsrd
